### PR TITLE
Add Option for Memory Optimized Training via Gradient Checkpointing

### PR DESCRIPTION
**Pull request by klei22**
#### Description:
This PR introduces gradient checkpointing to reduce memory usage during model training.

**Changes:**
- **Block Class**: Implemented gradient checkpointing using `torch.utils.checkpoint`, ensuring `use_reentrant=False` and `requires_grad=True` for inputs.
- **GPT Class**: Applied gradient checkpointing to each transformer block.

**Benefits:**
- **Memory Reduction**: Reduced GPU memory usage from around 19000MB to around 4000MB on the Shakespeare dataset with character tokenization, 512 context length and the softplus variation.
- **Feasibility**: Enables training larger models on existing hardware.

**Trade-offs:**
- **Increased Computation Time**: Activations are recomputed during the backward pass.
- **Added Complexity**: Additional complexity in the model implementation and training loop.

---

### Checklist:
- [x] Implemented gradient checkpointing in `Block` and `GPT` classes.
- [x] Tested for correct functionality and memory optimization.
- [x] Verified memory usage reduction on Shakespeare character dataset.
- [x] Updated documentation to reflect changes.
- [x] Ensured backward compatibility.

### CODE REVIEW

**Comment by klei22** [comment]
Wanted to note for _maximum_ memory savings, do not add the `--compile` flag.

For a context length of 1024, (gc will stand for 'gradient checkpointing')
- gc, no-compile -- 11074MB
- gc, compile -- 22846MB
- no-gc, no-compile -- OOM
- no-gc, compile -- OOM

File: model.py [hunk 1]
```diff
@@ -410,8 +420,22 @@ def forward(self, idx, targets=None):
...
+
+        x.requires_grad_(True)  # Ensure requires_grad is True
+
+        for block in self.transformer.h:
+            if self.config.use_gradient_checkpointing:
+                x = checkpoint.checkpoint(block, x, use_reentrant=False)
+            else:
+                x = block(x)
+
+        for block in self.transformer.h:
```
Comment (gkielian) [inline model.py:432]: Seems to be a duplicate of line 426

File: model.py [hunk 2]
```diff
@@ -410,8 +420,22 @@ def forward(self, idx, targets=None):
...
+            x = checkpoint.checkpoint(block, x, use_reentrant=False)
+
         x = self.transformer.ln_f(x)
 
+
```
Comment (gkielian) [inline model.py:437]: Let's remove these blank spaces

**Review by gkielian — CHANGES_REQUESTED** [review]

**Review by gkielian — APPROVED** [review]
